src/server/coprocessor/executor/scanner.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use server::coprocessor::Result;
+use kvproto::coprocessor::KeyRange;
+use storage::{Key, Value, SnapshotStore, Statistics, ScanMode};
+use util::escape;
+
+// `Scanner` is a helper struct to wrap all common scan operates
+// for `TableScanExecutor` and `IndexScanExecutor`
+pub struct Scanner<'a> {
+    store: SnapshotStore<'a>,
+    seek_key: Option<Vec<u8>>,
+    scan_mode: ScanMode,
+    upper_bound: Option<Vec<u8>>,
+    statistics: &'a mut Statistics,
+    desc: bool,
+    key_only: bool,
+}
+
+impl<'a> Scanner<'a> {
+    pub fn new(desc: bool,
+               key_only: bool,
+               store: SnapshotStore<'a>,
+               statistics: &'a mut Statistics)
+               -> Scanner<'a> {
+
+        let scan_mode = if desc {
+            ScanMode::Backward
+        } else {
+            ScanMode::Forward
+        };
+        Scanner {
+            store: store,
+            seek_key: None,
+            scan_mode: scan_mode,
+            upper_bound: None,
+            statistics: statistics,
+            desc: desc,
+            key_only: key_only,
+        }
+    }
+
+    pub fn get_row_from_range(&mut self, range: &KeyRange) -> Result<Option<(Vec<u8>, Value)>> {
+        let seek_key = self.prepare_and_get_seek_key(range);
+        if range.get_start() > range.get_end() {
+            return Ok(None);
+        }
+        let mut scanner = try!(self.store.scanner(self.scan_mode,
```

Comment:
Can `scanner` become a member variable? Now every time we perform `next`, we need to create a `scanner`.